### ISSUE TITLE
Parallel.lua local variable point to self.modules output reference, which means the outputs will only store calcuation result of final row/col for input

### ISSUE THREAD

**Issue by nomadshark** [torch/nn#402]
Parallel.lua, line 17:

---error code ----
function Parallel:updateOutput(input)
   local nModule=input:size(self.inputDimension)
   local outputs = {}

   for i=1,nModule do
      local currentInput = input:select(self.inputDimension,i)
      local currentOutput = self.modules[i]:updateOutput(currentInput)
      table.insert(outputs, currentOutput)
      local outputSize = currentOutput:size(self.outputDimension)
      
      if i == 1 then
         self.size:resize(currentOutput:dim()):copy(currentOutput:size())
      else
         self.size[self.outputDimension] = self.size[self.outputDimension] + outputSize
      end
      
   end
   self.output:resize(self.size)
   
   local offset = 1
   for i=1,nModule do
      local currentOutput = outputs[i]
      local outputSize = currentOutput:size(self.outputDimension)
      self.output:narrow(self.outputDimension, offset, outputSize):copy(currentOutput)
      offset = offset + currentOutput:size(self.outputDimension)
   end 
   return self.output
end

---correct code ----
function Parallel:updateOutput(input)
   local nModule=input:size(self.inputDimension)
   local outputs = {}

   for i=1,nModule do
      local currentInput = input:select(self.inputDimension,i)
      local currentOutput = self.modules[i]:updateOutput(currentInput):clone()
      table.insert(outputs, currentOutput)
      local outputSize = currentOutput:size(self.outputDimension)
      
      if i == 1 then
         self.size:resize(currentOutput:dim()):copy(currentOutput:size())
      else
         self.size[self.outputDimension] = self.size[self.outputDimension] + outputSize
      end
      
   end
   self.output:resize(self.size)
   
   local offset = 1
   for i=1,nModule do
      local currentOutput = outputs[i]
      local outputSize = currentOutput:size(self.outputDimension)
      self.output:narrow(self.outputDimension, offset, outputSize):copy(currentOutput)
      offset = offset + currentOutput:size(self.outputDimension)
   end 
   return self.output
end

**Comment by fmassa** [comment]
I don't think there need to be a `:clone` in https://github.com/torch/nn/blob/master/Parallel.lua#L17 . As `self.modules` is a table of modules, each individual module has its own `self.output`, which means that the `outputs` table will store the correct references.

**Comment by fmassa** [comment]
If you are facing some problems with`Parallel`, I suppose you are using a modules inside it that is in place. For example, `Identity` would fail, because it simply points to the input itself.

**Comment by fmassa** [comment]
Thinking a bit better, even if the modules inside `nn.Parallel` are in-place, you shouldn't face problems because of what I said earlier (`nn.Parallel` takes slices of the input as input to each module). But I might be missing something here, so could you provide an example that fails with current `nn.Parallel` ?

**Comment by fmassa** [comment]
So, I think the only situation in which you might have problems is when you are using in-place modules inside a `nn.Parallel`, and your input has zero stride in the dimension of the split. For example
```lua
m = nn.Parallel(1,1)
m:add(nn.AddConstant(1,true)) -- inplace
m:add(nn.AddConstant(2,true)) -- inplace
inputs = torch.ones(1,3):expand(2,3)
output = m:forward(inputs)
```
the output of this example is
```lua
th> output
 4
 4
 4
 4
 4
 4
[torch.DoubleTensor of size 6]
```
which is wrong. but if the modules are not in-place, everything works just fine.

**Comment by nomadshark** [comment]
:+1: 
thanks Francisco :) you're correct if modules are not in-place, it works just fine!
